SymbolCompletion shouldn't treat colon as a commit character in tuple

jcouv opened the pull request:
Not ready for review.

CyrusNajmabadi commented:
this PR was closed.  do you still want me to look at it?

jcouv commented:
That PR is not complete yet. It only handles symbols, but not keywords yet.
Do you think it's appropriate to create a SyntaxContext for keyword completion?

CyrusNajmabadi commented:
i don't understand.  KeywordProviders already have a syntax context.

jcouv commented:
Ah, you're right. It's just not available at the point were I need it. I'll figure it out. Thanks!

jcouv commented:
@CyrusNajmabadi The example you gave earlier (`case Color.Bl`) was good to illustrate that colon should remain a commit character for completing symbols in some cases. 
But I could not think of a similar example justifying colon as a commit character after a keyword. As it stands, this PR removes colon as a commit character from all keyword completion.

CyrusNajmabadi commented:
No :)

what about "def<colon>".  That better expand out to "default:".

jcouv commented:
Definitely. I'd looked through the keyword list and somehow missed that one. Thanks! :-)
I think the PR is ready for review now. Both symbol and keyword completion are aware of tuple context.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SuggestionModeCompletionProviderTests.cs:
```diff
@@ -391,7 +391,7 @@ static void Main(string[] args)
     }
 }";
-            await VerifyBuilderAsync(markup);
+            await VerifyNotBuilderAsync(markup);
```
don't we still want a builder here?  This could be a parenthesized lambda you're writign.  This seems like something regressed.

jcouv replied:
Traced this back to this [line/commit](https://github.com/dotnet/roslyn/commit/a3d8ab1b34eab45abf56e83178d3ddbe66a5c49f#diff-532cf6fcbe8d404c428ba6de669fd3b7R394) which incorrectly modified the test.

`Console.CancelKeyPress += new ConsoleCancelEventHandler((a$$` in the test above shows a builder.
But `Console.CancelKeyPress += new ConsoleCancelEventHandler(((a$$` should not.
This was hidden for a while by the introduction of a builder when in possible tuple context.

CyrusNajmabadi replied:
Ah ok.  Do we have a test for the first case?

jcouv replied:
Yes, the first case is just above (called `IncompleteParenthesizedLambdaArgument`).

CyrusNajmabadi replied:
ok cool.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SuggestionModeCompletionProviderTests.cs:
```diff
@@ -419,7 +419,7 @@ static void Main(string[] args)
     }
 }";
-            await VerifyBuilderAsync(markup);
+            await VerifyNotBuilderAsync(markup);
```
Can you verify a builder here if the type isn't 'var' but is instead a good Action type?

jcouv replied:
Sure. Added test for that.

CyrusNajmabadi commented on src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb:
```diff
@@ -291,8 +333,99 @@ class C
+
+                state.SendTypeChars("F")
+                Await state.WaitForAsynchronousOperationsAsync()
+                Await state.AssertSelectedCompletionItem(displayText:="Foo", isHardSelected:=True)
```
AssertSelectedCompletionItem already called WaitForAsyncOperation.

jcouv replied:
Simplified. Thanks

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs:
```diff
@@ -99,22 +99,73 @@ internal override bool IsInsertionTrigger(SourceText text, int characterPosition
-        private static readonly CompletionItemRules s_itemRules = CompletionItemRules.Default.
-            WithFilterCharacterRule(CharacterSetModificationRule.Create(CharacterSetModificationKind.Remove, '<')).
-            WithCommitCharacterRule(CharacterSetModificationRule.Create(CharacterSetModificationKind.Add, '<'));
+        // matrix of rules depending on contexts, with the following dimensions:
```
It would be *much* cleaner to just have a dictionary and key off of ... wait for it... a Tuple :)

jcouv replied:
Sure ;-)

CyrusNajmabadi commented on src/Features/Core/Portable/Completion/Providers/AbstractKeywordCompletionProvider.cs:
```diff
@@ -48,23 +49,26 @@ public int GetHashCode(RecommendedKeyword obj)
-                    (doc, ct) => RecommendKeywordsAsync(doc, position, ct),
+                    async (doc, ct) =>
+                    {
+                        var syntaxContext = await CreateContextAsync(doc, position, ct).ConfigureAwait(false);
```
The IDE prefers to avoid async lambdas.  can you make this a normal Task returning lambda, and have it call the async helper?

jcouv replied:
Done

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -1131,6 +1131,15 @@ public static bool IsParameterTypeContext(this SyntaxTree syntaxTree, int positi
+            var token = tokenOnLeftOfPosition;
+            token = token.GetPreviousTokenIfTouchingWord(position);
+
+            return token.IsKind(SyntaxKind.OpenParenToken, SyntaxKind.CommaToken);
```
This is too aggressive.  We should be checking that the open paren is specifically part of a tuple literal, or a parenthesized expression.  I'm not sure about the comma.  But it's likely that by that time, we can just see that hte comma is parented by a tuple literal.

Here's the case where the above code is bad (and we should have a command handler test for this).  Say the user writes:

```Foo(X<colon>``` or ```Foo(args, X<colon>```.

Based on your code, we would consider both of these cases to be a TupleElementName context.  but in both cases we want to complete out the colon.

jcouv replied:
If you type `(1`, there is no tuple literal in the syntax tree yet. 
Once you reach the first comma, then we're good though.

So I'm thinking to check whether the parent is an invocation, rather than check if it's a tuple literal. Then `Foo(X` will work properly.

CyrusNajmabadi replied:
> If you type (1, there is no tuple literal in the syntax tree yet. 

That's why i said: "or a parenthesized expression" :)

> So I'm thinking to check whether the parent is an invocation

That's not enough.  it doesn't cover things like ObjectCreationExpression.  Don't blacklist.  It's really hard to get right.  Instead, whitelist the cases we know we need to support.

CyrusNajmabadi commented:
Other cases that would be bad if you just check for open paren, or if it's not an invocation expression:

typeof
checked
default
casts
if/switch/while/for/foreach/catch/lock/using/
any parameter list.
any this/base list. 
attributes

jcouv commented:
@CyrusNajmabadi I think the latest commit addresses your feedback.
Btw, in terms of priority for the bug, do you think I should ask to get this in preview 5 (instead of master)?

jcouv commented:
Hum, thanks for the examples. I'm not sure what the heuristic for detecting tuple context should be... Until you get to the second element, the parser typically doesn't recognize there is a tuple (since it could be a parenthesized expression).

CyrusNajmabadi commented:
As mentioned, the algorithm should be as follows:

1. Look for the previous token (you're already doing this).
2. If it's a open paren and the paren is the start of a tuple or parenthesized expression, then this is a tuple context.
3. if it's a comma, and the parent of the comma is a tuple, then this is a tuple context.

Cheers!

jcouv commented:
On second thought, I don't see why it would make sense to have colon as a commit character for those examples you listed.

CyrusNajmabadi commented:
because we don't want to change our commit characters unless necessary.  that may break muscle memory.

CyrusNajmabadi commented:
For example, in an invocation expression, someone might do:  ```Foo(cancel<colon>``` because they want to complete out to ```cancellationToken:```

jcouv commented:
Thanks! I think that works.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs:
```diff
@@ -99,22 +99,69 @@ internal override bool IsInsertionTrigger(SourceText text, int characterPosition
+        protected override CompletionItemRules GetCompletionItemRules(List<ISymbol> symbols, SyntaxContext context, bool preselect)
+        {
+            return cachedRules[Tuple.Create(context.IsInImportsDirective, preselect,
+                ((CSharpSyntaxContext)context).IsPossibleTupleContext)] ??
```
I would just move IsPossibleTupleContext up to SyntaxContext.

jcouv replied:
Done

jcouv replied:
Only trouble is that those rules are not yet implemented for VB. I'll just pass false for now.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs:
```diff
@@ -99,22 +99,69 @@ internal override bool IsInsertionTrigger(SourceText text, int characterPosition
+                    {
+                        if (importDirective == 1 && tupleLiteral == 1)
+                        {
+                            // this combination doesn't make sense, we can skip it
```
THe level of micro-optimization here is pretty impressive :)

CyrusNajmabadi replied:
i would remove it to make the core lookup code simple.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs:
```diff
@@ -99,22 +99,69 @@ internal override bool IsInsertionTrigger(SourceText text, int characterPosition
+        private static readonly Dictionary<Tuple<bool, bool, bool>, CompletionItemRules> cachedRules = InitCachedRules();
 
-        private static readonly CompletionItemRules s_itemRules_preselect = s_itemRules.WithSelectionBehavior(CompletionItemSelectionBehavior.HardSelection);
+        private static Dictionary<Tuple<bool, bool, bool>, CompletionItemRules> InitCachedRules()
```
ValueTuple.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs:
```diff
@@ -99,22 +99,69 @@ internal override bool IsInsertionTrigger(SourceText text, int characterPosition
-        private static CompletionItemRules s_importDirectiveRules_preselect = s_importDirectiveRules.WithSelectionBehavior(CompletionItemSelectionBehavior.HardSelection);
+        protected override CompletionItemRules GetCompletionItemRules(List<ISymbol> symbols, SyntaxContext context, bool preselect)
+        {
+            return cachedRules[Tuple.Create(context.IsInImportsDirective, preselect,
```
ValueTuple so we don't allocate.

jcouv replied:
Done

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -1131,6 +1131,16 @@ public static bool IsParameterTypeContext(this SyntaxTree syntaxTree, int positi
+            var token = tokenOnLeftOfPosition;
+            token = token.GetPreviousTokenIfTouchingWord(position);
+
+            return (token.IsKind(SyntaxKind.OpenParenToken) && token.Parent.IsKind(SyntaxKind.ParenthesizedExpression)) ||
```
for the open paren case, you also need to check if the parent is a tuple.  You should also have a test for this.  i.e. completion when the user has:   ```(a<colon>,```.  Note that the comma is already there.

CyrusNajmabadi replied:
i.e. 

```
if (token is OpenParen) return parent is Parenthesized or Tuple
if (token is Comma) return parent is Tuple
```

jcouv replied:
Done

jcouv commented:
Pushed the fixes we discussed. Also, please advise on preview 5 vs. master. Thanks!

CyrusNajmabadi commented:
@Pilchie  can speak to preview 5.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/SymbolCompletionProvider.cs:
```diff
@@ -99,22 +99,69 @@ internal override bool IsInsertionTrigger(SourceText text, int characterPosition
+        protected override CompletionItemRules GetCompletionItemRules(List<ISymbol> symbols, SyntaxContext context, bool preselect)
+        {
+            return cachedRules[ValueTuple.Create(context.IsInImportsDirective, preselect,
+                ((CSharpSyntaxContext)context).IsPossibleTupleContext)] ??
```
Cast should not be needed if this moved up into SyntaxContext.

CyrusNajmabadi replied:
Indexing into an array will throw if the item is not in the dictionary.  you either need TryGetValue, or you remove the below optimization and you just include all 8 values.

jcouv replied:
Removed cast and fixed bug (using TryGetValue).

CyrusNajmabadi replied:
The cast still seems to be there.

CyrusNajmabadi replied:
it doesn't look like you're calling TryGetValue.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/SuggestionMode/CSharpSuggestionModeCompletionProvider.cs:
```diff
@@ -143,6 +125,13 @@ private bool IsLambdaExpression(SemanticModel semanticModel, int position, Synta
                 }
             }
 
+            // A lambda that is being typed may be parsed as a tuple without names 
```
give an example.   i.e. "(a, b"  could be the start of a tuple/lambda, but "(a: b, c" could never be a lambda.

CyrusNajmabadi replied:
This comments seems to be the same.

CyrusNajmabadi commented:
LGTM if the buggy dictionary code is fixed.

jcouv commented:
I see that Jared pushed this to RC milestone. I'll go ahead with merging to master.
